File: sample.R
#Computation for HBV caccination data
#methylation
sample <- read.csv("sample.csv",sep="\t",header=F)
index <- order(as.numeric(gsub("_V","",gsub("GR","",sample[,2]))))
x_methyl <- read.csv("GSE161020_series_matrix.txt.gz",sep="\t",comment.char="!")
X <-  t(data.matrix(scale(x_methyl[,-1][,index]))) %*% data.matrix(scale(x_methyl[,-1][,index]))
#Gene expression
files <- list.files("./",pattern="count.txt.gz")
x_all <- NULL
for (i in c(1:length(files)))
{
    cat(i," ")
    x <- read.csv(files[i],sep="\t",header=F)
    x_all <- cbind(x_all,x[,2])
}
x_all <- data.frame(x[,1],x_all)
X0 <- t(data.matrix(scale(x_all[,-1]))) %*% data.matrix(scale(x_all[,-1]))
#proteome
x <- read.csv("GR01,04,09,10,11,13,15,17,18,19.txt",sep="\t")
x1 <- read.csv("GR02,03,05,06,07.txt",sep="\t")
x[1:3,2] <- c("a","b","c")
x1[1:3,2] <- c("a","b","c")
x0 <- merge(x,x1,by.x=2,by.y=2,all=T,sort=F)
cells <- unlist(lapply(strsplit(colnames(x0),".",fixed=T),"[",1))
GR <- unlist(x0[1,])
visit <- unlist(x0[2,])
labels_cells <- names(table(cells))[-3]
labels_GR<- names(table(GR))[-c(1:2)]
labels_visit <- names(table(visit))[-c(1:2)]
visit[!(visit %in% labels_visit)] <- "X"
cells[!(cells %in% labels_cells)] <- "X"
GR[!(GR %in% labels_GR)] <- "X"


x_all_WBC <- NULL
x_all_Plasma <- NULL

for (i in c(1:length(labels_GR)))
{
    cat(i," ")
    for (j in c(1:length(labels_visit)))
    {
        index_WBC <- visit==labels_visit[j] & GR==labels_GR[i] & cells=="WBC"
        x_all_WBC = cbind(x_all_WBC,x0[,index_WBC,drop=F][,1])
        index_plasma <- visit==labels_visit[j] & GR==labels_GR[i] & cells=="Plasma"
        x_all_Plasma = cbind(x_all_Plasma,x0[,index_plasma,drop=F][,1])
    }
}
x_all_WBC <- data.frame(x0[,1],x_all_WBC)
x_all_Plasma <- data.frame(x0[,1],x_all_Plasma)
x_all_WBC[is.na(x_all_WBC)]<-0
x_all_Plasma[is.na(x_all_Plasma)]<-0
x_all_WBC[,-1] <- apply(x_all_WBC[,-1],2,as.numeric)
x_all_WBC[is.na(x_all_WBC)]<-0
x_all_WBC[,-1] <- scale(x_all_WBC[,-1])
x_all_WBC[is.na(x_all_WBC)]<-0
X1_WBC <- t(data.matrix(x_all_WBC[,-1])) %*% data.matrix(x_all_WBC[,-1])
x_all_Plasma[,-1] <- apply(x_all_Plasma[,-1],2,as.numeric)
x_all_Plasma[is.na(x_all_Plasma)]<-0
x_all_Plasma[,-1] <- scale(x_all_Plasma[,-1])
x_all_Plasma[is.na(x_all_Plasma)]<-0
X1_Plasma <- t(data.matrix(x_all_Plasma[,-1])) %*% data.matrix(x_all_Plasma[,-1])
dim(X) <- c(5,15,75)
dim(X0) <- c(5,15,75)
dim(X1_WBC) <- c(5,15,75)
dim(X1_Plasma) <- c(5,15,75)
Z <- array(NA,c(75,75,4))
Z[,,1] <- X/mean(X)
Z[,,2] <- X0/mean(X0)
Z[,,3] <- X1_WBC/mean(X1_WBC)
Z[,,4] <- X1_Plasma/mean(X1_Plasma)
dim(Z) <- c(5,15,5,15,4)
require(rTensor)
HOSVD <- hosvd(as.tensor(Z))
u <- outer(HOSVD$U[[1]][,2],HOSVD$U[[2]][,1])
dim(u) <- 75
P <- pchisq(scale(data.matrix(scale(x_methyl[,-1][,index])) %*% u)^2,1,lower.tail=F)
table(p.adjust(P,"BH")<0.01)
y <- read.csv("GPL21145_MethylationEPIC_15073387_v-1-0.csv.gz",sep=",",skip=7)
gene <- y[match(x[p.adjust(P,"BH")<0.01,1],y[,1]),15]
gene <- sort(unique(unlist(strsplit(gene,";"))))
P <- pchisq(scale(data.matrix(scale(x_all[,-1])) %*% u)^2,1,lower.tail=F)
gene <- x_all[p.adjust(P,"BH")<0.01,1] #genes selected by mnethylation
P <- pchisq(scale(data.matrix(x_all_WBC[,-1]) %*% u)^2,1,lower.tail=F)
gene <- x_all[p.adjust(P,"BH")<0.01,1] #genes selected by gene expression
gene <- unlist(strsplit(x_all_WBC[p.adjust(P,"BH")<0.05,1],"|",fixed=T)) 
gene <- gene[grep("NP",gene)] 
gene <- unlist(lapply(strsplit(gene,".",fixed=T),"[",1)) #genes selected by WBC
P <- pchisq(scale(data.matrix(x_all_Plasma[,-1]) %*% scale(u))^2,1,lower.tail=F)
gene <- unlist(strsplit(x_all_Plasma[p.adjust(P,"BH")<0.05,1],"|",fixed=T))
gene <- gene[grep("NP",gene)]
gene <- unlist(lapply(strsplit(gene,".",fixed=T),"[",1)) #genes selected by Plasma

#Synthetic data set
label <- matrix(F,ncol=3,nrow=1000)
  label[1:10,]<-T
  for (j in c(1:3))
  {
      label[10*j+c(1:10),j] <-T
  }
 P1 <- NULL
 require(rTensor)
 xx <- seq(0,1,length=10)
for (k in c(1:100))
{
    cat(k," ")
  x <- matrix(runif(1000*10),10,1000)
  x1 <- matrix(runif(1000*10),10,1000)
  x2 <- matrix(runif(1000*10),10,1000)
  x[,1:10] <- x[,1:10]+xx
  x1[,1:10] <- x1[,1:10]+xx
  x2[,1:10] <- x2[,1:10]+xx
  x[,11:20] <- x[,11:20]+xx
  x1[,21:30] <- x1[,21:30]+xx
  x2[,31:40] <- x2[,31:40]+xx
  x <- t(x)
  x1 <- t(x1)
  x2 <- t(x2)
  Z <- array(NA,c(1000,10,3))
  Z[,,1] <- x
  Z[,,2] <- x1
  Z[,,3] <- x2
  dim(Z) <- c(1000,30)
  ZZ <- t(Z) %*% Z
  dim(ZZ) <- c(10,3,10,3)
  HOSVD <- hosvd(as.tensor(ZZ))
  plot(HOSVD$U[[1]][,2])
  plot(HOSVD$U[[2]][,1])
  U <- Z %*% as.vector(outer(HOSVD$U[[1]][,2],HOSVD$U[[2]][,1]))
  plot(U)
  P <- pchisq(scale(U)^2,1,lower.tail=F)
  P1 <- cbind(P1,P)
  }
  
 PP1 <- apply(P1,2,function(x){p.adjust(x,"BH")})
 sum(rowSums(PP1<0.01)[1:10])
 
 sum(rowSums(PP1<0.01)[11:1000])

File: sample.py
import pandas as pd
import numpy as np
import scipy.stats
import glob
import collections


# in order to reorder methylation samples
sample = pd.read_csv("sample.csv", sep="\t", header=None)


def remove(s):
    return float(s.replace('GR', '').replace('_V', ''))


a = list(map(remove, sample.values[:, 1]))
index = np.argsort(a)

# methylation
x = pd.read_csv("GSE161020_series_matrix.txt.gz", sep="\t", comment="!")
y = scipy.stats.zscore(x.values[:, 1:][:, index].astype(np.float64))
np.save("y_methyl",y)
X = y.T@y
np.save("X",X)

# gene expression
path = './*.count.txt.gz'
files = glob.glob(path)
files = sorted(files)
for i in range(len(files)):
    print(i)
    x = pd.read_csv(files[i], sep="\t", header=None)
    if i == 0:
        x_all = x.values[:, 1]
    else:
        x_all = np.vstack((x_all, x.values[:, 1]))

x_all = x_all.T
y = scipy.stats.zscore(x_all.astype(np.float64))
np.save("y_exp",y)
X0 = y.T@y
np.save("X0",X0)

# proteome
x = pd.read_csv("GR01,04,09,10,11,13,15,17,18,19.txt", sep="\t")
x1 = pd.read_csv("GR02,03,05,06,07.txt", sep="\t")
x.values[0:3, 1] = ["a", "b", "c"]
x1.values[0:3, 1] = ["a", "b", "c"]
x0 = pd.merge(x, x1, on='Unnamed: 1', how='outer')
cells = np.array([l[0] for l in x0.columns.str.split(".")])
GR = x0.values[0, ]
visit = x0.values[1, ]
np.save("x0",x0)

def remove2(s):
    return s.replace('_x', '').replace('_y', '')


cells = np.array(list(map(remove2, cells)))
labels_cells = [label for label in collections.Counter(cells).keys()]
labels_cells = sorted(np.array(labels_cells))
labels_cells = [labels_cells[i] for i in [0, 3]]
labels_GR = [label for label in collections.Counter(GR).keys()]
labels_GR = sorted(np.array(labels_GR))
labels_GR = labels_GR[:-2]
labels_visit = [label for label in collections.Counter(visit).keys()]
labels_visit = sorted(np.array(labels_visit))
labels_visit = labels_visit[:-2]
# use np.isin
visit[np.bitwise_not(np.isin(visit, labels_visit))] = "X"
cells[np.bitwise_not(np.isin(cells, labels_cells))] = "X"
GR[np.bitwise_not(np.isin(GR, labels_GR))] = "X"

#visit[[i not in labels_visit for i in visit]] = "X"
#cells[[i not in labels_cells for i in cells]] = "X"
#GR[[i not in labels_GR for i in GR]] = "X"

id_WBC = None
id_Plasma = None
for j in range(len(labels_GR)):
    for i in range(len(labels_visit)):        
        index = [x == labels_visit[i] and y == labels_GR[j] for (x, y) in zip(visit, GR)]
        index_WBC = [x and y == "WBC" for (x, y) in zip(index, cells)]
        index_Plasma = [x and y == "Plasma" for (x, y) in zip(index, cells)]
        id_WBC = np.hstack((id_WBC, np.arange(183)[index_WBC][0]))
        id_Plasma = np.hstack((id_Plasma, np.arange(183)[index_Plasma][0]))
       # print(i,j,np.arange(183)[index_WBC],np.arange(183)[index_Plasma])

id_WBC = id_WBC[1:]
id_Plasma = id_Plasma[1:]


x0_WBC = x0.values[:, list(map(int, id_WBC))]
x0_Plasma = x0.values[:, list(map(int, id_Plasma))]

x0_WBC = x0_WBC[4:, :].astype(np.float64)
x0_Plasma = x0_Plasma[4:, :].astype(np.float64)

x0_WBC = np.nan_to_num(x0_WBC)
x0_Plasma = np.nan_to_num(x0_Plasma)

y = scipy.stats.zscore(x0_WBC)
np.save("y_WBC",y)
X1_WBC = y.T@y

y = scipy.stats.zscore(x0_Plasma)
np.save("y_Plasma",y)
X1_Plasma = y.T@y

np.save("X1_WBC",X1_WBC)
np.save("X1_Plasma",X1_Plasma)

#---- 
X = np.load("X.npy")
X0 = np.load("X0.npy")
X1_WBC = np.load("X1_WBC.npy")
X1_Plasma = np.load("X1_Plasma.npy")
X = np.reshape(X,[15,5,15,5])
X = X.T
X = X/np.mean(X)
X0 = np.reshape(X0,[15,5,15,5])
X0 = X0.T
X0 = X0/np.mean(X0)
X1_WBC = np.reshape(X1_WBC,[15,5,15,5])
X1_WBC = X1_WBC.T
X1_WBC = X1_WBC/np.mean(X1_WBC)
X1_Plasma = np.reshape(X1_Plasma,[15,5,15,5])
X1_Plasma = X1_Plasma.T
X1_Plasma = X1_Plasma/np.mean(X1_Plasma)

import tensorly as tl

Z = tl.tensor(np.zeros(75*75*4).reshape(5,15,5,15,4))
Z[:,:,:,:,0]=X
Z[:,:,:,:,1]=X0
Z[:,:,:,:,2]=X1_WBC
Z[:,:,:,:,3]=X1_Plasma
from tensorly.decomposition import tucker
core, factors = tucker(Z, rank=[5,15,5,15,4],n_iter_max=1,init='svd')

np.save("core",core)
np.save("factors",factors)

#-----
import statsmodels.stats.multitest

core = np.load("core.npy")
factors = np.load("factors.npy",allow_pickle=True)
u= np.outer(factors[0][:,1],factors[1][:,0])
u.reshape(75)


#methylation
y = np.load("y_methyl.npy")
P = scipy.stats.chi2.sf(scipy.stats.zscore(scipy.stats.zscore(y)@u.T.reshape(75))**2,1)
P0 = np.zeros(len(P))
P0.fill(0.01)
collections.Counter(statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0])

#Counter({False: 685505, True: 2077})

y = pd.read_csv("GPL21145_MethylationEPIC_15073387_v-1-0.csv.gz",sep=",",skiprows=7)
gene = y.values[:,14][np.isin(y.values[:,0],x.values[statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0],0])]

#gene expression
y=np.load("y_exp.npy")
P = scipy.stats.chi2.sf(scipy.stats.zscore(scipy.stats.zscore(y)@u.T.reshape(75))**2,1)
P0 = np.zeros(len(P))
P0.fill(0.01)
collections.Counter(statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0])

#Counter({False: 58296, True: 11})
x = pd.read_csv(files[0], sep="\t", header=None)
gene = x.values[statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0],0]

#Proteome 
#WBC
y=np.load("y_WBC.npy")
P = scipy.stats.chi2.sf(scipy.stats.zscore(scipy.stats.zscore(y)@u.T.reshape(75))**2,1)
P0 = np.zeros(len(P))
P0.fill(0.05)
collections.Counter(statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0])

#Counter({False: 1579, True: 5})

x0=np.load("x0.npy",allow_pickle=True)
gene = x0[4:,0][statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0]]

#Plasma
y=np.load("y_Plasma.npy")
P = scipy.stats.chi2.sf(scipy.stats.zscore(scipy.stats.zscore(y)@u.T.reshape(75))**2,1)
P0 = np.zeros(len(P))
P0.fill(0.05)
collections.Counter(statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0])

#Counter({False: 1579, True: 5})

gene = x0[4:,0][statsmodels.stats.multitest.multipletests(P,0.01,method="fdr_bh")[0]]
